src/ApacheOrcDotNet.OptimizedReader/Buffers/BaseColumnBuffer.cs

```diff
@@ -55,100 +44,89 @@ private protected void ReadByteStream(StreamDetail stream, ReadOnlySpan<byte> de
-                        return;
+                    rleValues[0] = (decodedByte & 128) != 0;
+                    rleValues[1] = (decodedByte & 64) != 0;
+                    rleValues[2] = (decodedByte & 32) != 0;
+                    rleValues[3] = (decodedByte & 16) != 0;
+                    rleValues[4] = (decodedByte & 8) != 0;
+                    rleValues[5] = (decodedByte & 4) != 0;
+                    rleValues[6] = (decodedByte & 2) != 0;
+                    rleValues[7] = (decodedByte & 1) != 0;
 
-                    outputValues[numValuesRead++] = (decodedByte & 64) != 0;
-                    if (checkRemainingBits && ++numSkippedBits <= stream.Positions.RemainingBits)
-                        numValuesRead--;
-                    if (numValuesRead >= outputValues.Length)
-                        return;
+                    if (isFirstByte)
+                    {
+                        isFirstByte = false;
 
-                    outputValues[numValuesRead++] = (decodedByte & 32) != 0;
-                    if (checkRemainingBits && ++numSkippedBits <= stream.Positions.RemainingBits)
-                        numValuesRead--;
-                    if (numValuesRead >= outputValues.Length)
-                        return;
+                        var firstValues = rleValues[stream.Positions.RemainingBits..];
 
-                    outputValues[numValuesRead++] = (decodedByte & 16) != 0;
-                    if (checkRemainingBits && ++numSkippedBits <= stream.Positions.RemainingBits)
-                        numValuesRead--;
-                    if (numValuesRead >= outputValues.Length)
-                        return;
+                        numValuesRead += firstValues.Length;
 
-                    outputValues[numValuesRead++] = (decodedByte & 8) != 0;
-                    if (checkRemainingBits && ++numSkippedBits <= stream.Positions.RemainingBits)
-                        numValuesRead--;
-                    if (numValuesRead >= outputValues.Length)
-                        return;
-
-                    outputValues[numValuesRead++] = (decodedByte & 4) != 0;
-                    if (checkRemainingBits && ++numSkippedBits <= stream.Positions.RemainingBits)
-                        numValuesRead--;
-                    if (numValuesRead >= outputValues.Length)
-                        return;
+                        firstValues.CopyTo(outputValues);
+                    }
+                    else
+                    {
+                        var targetBuffer = outputValues[numValuesRead..];
 
-                    outputValues[numValuesRead++] = (decodedByte & 2) != 0;
-                    if (checkRemainingBits && ++numSkippedBits <= stream.Positions.RemainingBits)
-                        numValuesRead--;
-                    if (numValuesRead >= outputValues.Length)
-                        return;
+                        if (numValuesRead + rleValues.Length >= outputValues.Length)
```

Comment:
Add a comment that this is considered only on the last byte.  The above is clear with `if (isFirstByte)`.

Reply:
Cleaned up the code and added comments. One observation here is that this block will execute when the number of read values will complete or exceed the expected number of output values, and that can happen before the last byte is reached.